### ISSUE TITLE
Detect hostname at Vector start time

### ISSUE THREAD

**Issue by jszwedko** [vectordotdev/vector#14250]
Vector currently fetches the hostname when needed at runtime:

https://github.com/vectordotdev/vector/blob/05fc485f52eba72eb4bb68ab6d392753138f0f2d/src/sources/internal_metrics.rs#L124

If this fails, behavior is usually degraded. In the above example, the internal metrics end up being published without a hostname. As a user, I think I'd prefer that the metrics not be published than be published without a hostname.

I think we could mitigate this by:

* Fetching the hostname at start time, failing the Vector boot if it cannot be fetched
* Still fetch the hostname at runtime as needed, in case it has been updated after Vector starts, but if it cannot be fetched, fall back to the cached value

**Comment by spencergilbert** [comment]
Caveats - the hostname may be the Pod's name if Vector is running in kubernetes. I'd expect similar behavior would be possible in other containerized situations.

**Comment by jszwedko** [comment]
> Caveats - the hostname may be the Pod's name if Vector is running in kubernetes. I'd expect similar behavior would be possible in other containerized situations.

Yeah, agreed. I'm not proposing any changes to the hostname detection mechanism here (which for now is just essentially `gethostname`), but that is something we may want to think about in the future.

**Comment by spencergilbert** [comment]
Related https://github.com/vectordotdev/vector/pull/13703
